src/infill/LightningLayer.cpp

```diff
@@ -0,0 +1,365 @@
+//Copyright (c) 2021 Ultimaker B.V.
+//CuraEngine is released under the terms of the AGPLv3 or higher.
+
+#include "LightningLayer.h"
+
+#include <iterator> // advance
+
+#include "LightningTree.h"
+
+#include "../sliceDataStorage.h"
+#include "../utils/linearAlg2D.h"
+#include "../utils/SVG.h"
+#include "../utils/SparsePointGridInclusive.h"
+
+using namespace cura;
+
+coord_t LightningLayer::getWeightedDistance(const Point& boundary_loc, const Point& unsupported_loc)
+{
+    return vSize(boundary_loc - unsupported_loc);
+}
+
+LightningDistanceField::LightningDistanceField
+(
+    const coord_t& radius,
+ const Polygons& current_outline,
+ const Polygons& current_overhang,
+ const std::vector<std::shared_ptr<LightningTreeNode>>& initial_trees
+)
+: cell_size(radius / 6) // TODO: make configurable!
+, grid(cell_size)
+, supporting_radius(radius)
+, current_outline(current_outline)
+, current_overhang(current_overhang)
+{
+    std::vector<Point> regular_dots = PolygonUtils::spreadDotsArea(current_overhang, cell_size);
+    for (Point p : regular_dots)
+    {
+        const ClosestPolygonPoint cpp = PolygonUtils::findClosest(p, current_outline);
+        const coord_t dist_to_boundary = vSize(p - cpp.p());
+        unsupported_points.emplace_back(p, dist_to_boundary);
+    }
+    unsupported_points.sort(
+        [](const UnsupCell& a, const UnsupCell& b)
+        {
+            coord_t da = a.dist_to_boundary + std::hash<Point>{}(a.loc) % 191;
+            coord_t db = b.dist_to_boundary + std::hash<Point>{}(b.loc) % 191;
+            return da < db;
+        });
+    for (auto it = unsupported_points.begin(); it != unsupported_points.end(); ++it)
```

Comment:
... because we place the iterator `it` into the `unsupported_point_grid`. Would not have been my first choice. If I've got a bit of time left I can have a look if I can rewrite this with shared or weak pointers.